### PR TITLE
haskell.packages.*.cabal-install*: unify overrides, haskellPackages.guardian: fix build

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
@ofborg build haskell.packages.ghc8107.cabal-install haskell.packages.ghc884.cabal-install haskell.packages.ghc902.cabal-install haskell.packages.ghc924.cabal-install haskell.packages.ghc925.cabal-install haskell.packages.ghc926.cabal-install haskell.packages.ghc944.cabal-install haskell.packages.ghc8107.cabal-install-solver haskell.packages.ghc884.cabal-install-solver haskell.packages.ghc902.cabal-install-solver haskell.packages.ghc924.cabal-install-solver haskell.packages.ghc925.cabal-install-solver haskell.packages.ghc926.cabal-install-solver haskell.packages.ghc944.cabal-install-solver haskellPackages.guardian

File: pkgs/development/haskell-modules/configuration-nix.nix [hunk 1]
```diff
@@ -1053,10 +1053,12 @@ self: super: builtins.intersectAttrs super {
...
 
   # Make sure that Cabal 3.8.* can be built as-is
-  Cabal_3_8_1_0 = doDistribute (super.Cabal_3_8_1_0.override {
+  Cabal_3_8_1_0 = doDistribute (super.Cabal_3_8_1_0.override ({
     Cabal-syntax = self.Cabal-syntax_3_8_1_0;
+  } // lib.optionalAttrs (lib.versionOlder self.ghc.version "9.2.5") {
+    # Use process core package when possible
     process = self.process_1_6_16_0;
-  });
+  }));
```
Comment (maralorn) [inline pkgs/development/haskell-modules/configuration-nix.nix:1061]: Is there a good reason to have this override in configuration-nix?

**Review by maralorn — APPROVED** [review]
This looks pretty ingenious. Well done!
Code looks sensible and apparently it builds.

File: pkgs/development/haskell-modules/configuration-nix.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/haskell-modules/configuration-nix.nix:1061]: No particular reason, just was there before. I guess these kinds of overrides are a gray area.

Comment (maralorn) [reply pkgs/development/haskell-modules/configuration-nix.nix:1061]: I agree. I’m fine either way.

**Pull request merged**